### ISSUE TITLE
Take ()->FlxState in FlxG.switchState

### ISSUE THREAD

**Issue by Geokureli** [HaxeFlixel/flixel#2541]
In it's current form: `FlxG.switchState(new PlayState())` the state's constructor is called immediately, and the stateSwitch happens later, hence why we have `create()`. I'm proposing this should take a function that returns a newly created FlxState instance. This would allow devs to put all of their init logic in the constructor, where as of now this pattern would cause issues, since the constructor is called while the old state is still active.

This will also allow us to call `FlxG.resetState` with states that have constructor args

Similarly, the initialState passed in the FlxGame constructor should also be changed to a function.

## Easy migration guide
```hx
// old code                                  -->    new code
FlxG.switchState(new PlayState());        /* --> */ FlxG.switchState(PlayState.new);
FlxG.switchState(new PlayState(levelID)); /* --> */ FlxG.switchState(PlayState.new.bind(levelID));
```

~~This is literally gonna break every demo~~ Edit: IT DIDN'T!
This is still a breaking change, becuase it requires the removal of `FlxState.switchTo`

Recommended by @Gama11

**Comment by Cheemsandfriends** [comment]
Got a question, will this have backwards thingy?
like, two options or just like make this as definitive?

**Comment by Geokureli** [comment]
no backward compatibility, this is a breaking change and it's gonna fuck shit up. but it's better in the long run

**Comment by Cheemsandfriends** [comment]
> Also I plan to release 5.0.0beta with everything except this change, so people can ease into the changes.

how about in the beta make a tweening on the old and new way, so that they can use the old but get deprecations and the new one, and on final release be like, boom gone

**Comment by Geokureli** [comment]
I don't know what tweening means, in this context, I don't know how i would make this backward compatible without doing some hacky stuff breaking typeSafety, and I do think forcing migration is better than silently changing everything

**Comment by Cheemsandfriends** [comment]
> I don't know what tweening means, in this context, I don't know how i would make this backward compatible without doing some hacky stuff breaking typeSafety, and I do think forcing migration is better than silently changing everything

basically like, making a backwards compatibility for the beta, but it will be a deprecated warning like warning that in the final release, it will change, or something like that

**Comment by Geokureli** [comment]
do you have a suggestion? I can't think of a way to accomplish that

**Comment by Cheemsandfriends** [comment]
> do you have a suggestion? I can't think of a way to accomplish that

I'm thinking of a way but it would be runtime, not compilation warning tbh

**Comment by Cheemsandfriends** [comment]
@Geokureli I need to ask this just in case, the `() -> FlxState` is supposed to make `create()` deprecated or something like that?
cos I've been trying to add them together but it's been hard to make them work together seamlessly

**Comment by Geokureli** [comment]
I imagine the flow will be 
- set FlxG.state to null and destroy the previous state
- instantatiate the new state
- set FlxG.state to the new state
- call state.create()

for 99% of the cases people can put all init logic in the state's constructor, but if they depend on a global reference to the state for some reason they can use create()

**Comment by BeastlyGabi** [comment]
came here from [this](https://github.com/HaxeFlixel/flixel/pull/2571), I'm __assuming__ with this breaking change error log will be objectively better than it is now? just curious, that's all!

**Comment by Geokureli** [comment]
> came here from [this](https://github.com/HaxeFlixel/flixel/pull/2571), I'm **assuming** with this breaking change, error logging will be objectively better than it is now? just curious, that's all!

sorry for the late reply, this will prevent errors caused when instantiating flixel objects in a FlxState's constructor, or referencing global flixel tools in a FlxState's constructor. this will effectively remove the need for the `create` method which will open up the door for other possibilities.

**Comment by Geokureli** [comment]
closed via #2733
